### Note
* Instructions have been included for each segment. You do not have to follow them exactly, but they are included to help you think through the steps.

In [1]:
# Dependencies and Setup
import pandas as pd

# File to Load (Remember to Change These)
school_data_to_load  = "Resources/schools_complete.csv"
student_data_to_load = "Resources/students_complete.csv"

# Read School and Student Data File and store into Pandas DataFrames
school_data = pd.read_csv(school_data_to_load)
student_data = pd.read_csv(student_data_to_load)

# Combine the data into a single dataset.  
school_data_complete = pd.merge(student_data, school_data, how="left", on=["school_name", "school_name"])
school_data_complete

Student ID       student_name gender grade         school_name  \
0               0       Paul Bradley      M   9th   Huang High School   
1               1       Victor Smith      M  12th   Huang High School   
2               2    Kevin Rodriguez      M  12th   Huang High School   
3               3  Dr. Richard Scott      M  12th   Huang High School   
4               4         Bonnie Ray      F   9th   Huang High School   
...           ...                ...    ...   ...                 ...   
39165       39165       Donna Howard      F  12th  Thomas High School   
39166       39166          Dawn Bell      F  10th  Thomas High School   
39167       39167     Rebecca Tanner      F   9th  Thomas High School   
39168       39168       Desiree Kidd      F  10th  Thomas High School   
39169       39169    Carolyn Jackson      F  11th  Thomas High School   

       reading_score  math_score  School ID      type  size   budget  
0                 66          79          0  District  2917  1910635  
1                 94          61          0  District  2917  1910635  
2                 90          60          0  District  2917  1910635  
3                 67          58          0  District  2917  1910635  
4                 97          84          0  District  2917  1910635  
...              ...         ...        ...       ...   ...      ...  
39165             99          90         14   Charter  1635  1043130  
39166             95          70         14   Charter  1635  1043130  
39167             73          84         14   Charter  1635  1043130  
39168             99          90         14   Charter  1635  1043130  
39169             95          75         14   Charter  1635  1043130  

[39170 rows x 11 columns]

## District Summary

* Calculate the total number of schools

* Calculate the total number of students

* Calculate the total budget

* Calculate the average math score 

* Calculate the average reading score

* Calculate the percentage of students with a passing math score (70 or greater)

* Calculate the percentage of students with a passing reading score (70 or greater)

* Calculate the percentage of students who passed math **and** reading (% Overall Passing)

* Create a dataframe to hold the above results

* Optional: give the displayed data cleaner formatting

In [2]:
# Total number of schools
total_schools = school_data_complete["school_name"].nunique()

In [3]:
# Total number of students
total_students = school_data_complete["Student ID"].count()

In [4]:
# Total budget
total_budget = sum(school_data_complete["budget"].unique())

In [5]:
# Average math score
average_math = school_data_complete["math_score"].mean()

In [6]:
# Average reading score
average_reading = school_data_complete["reading_score"].mean()

In [7]:
# Percentage of students with passing math score 70 or higher
passing_math = (school_data_complete["math_score"] >= 70).mean() * 100

In [8]:
# Percentage of students with passing reading score 70 or higher
passing_reading = (school_data_complete["reading_score"] >= 70).mean() * 100

In [9]:
# Percentage of students who passed math and reading(% Overall Passing)
passing_readandmath = ((school_data_complete["reading_score"] >= 70) & (school_data_complete["math_score"] >= 70)).mean() * 100

In [10]:
# Creating district_summary_df
district_summary_df = pd.DataFrame({"Total Schools":[total_schools],
                                    "Total Students":[total_students],
                                    "Total Budget":[total_budget],
                                    "Average Math Score":[average_math],
                                    "Average Reading Score":[average_reading],
                                    "% Passing Math":[passing_math],
                                    "% Passing Reading":[passing_reading],
                                    "% Overall Passing":[passing_readandmath]
                                   })

In [11]:
# Cleaning up formatting with .map
district_summary_df["Total Students"] = district_summary_df["Total Students"].map("{:,}".format)
district_summary_df["Total Budget"] = district_summary_df["Total Budget"].map("${:,.2f}".format)
district_summary_df["Average Math Score"] = district_summary_df["Average Math Score"].map("{:.6f}".format)
district_summary_df["Average Reading Score"] = district_summary_df["Average Reading Score"].map("{:.5f}".format)
district_summary_df["% Passing Math"] = district_summary_df["% Passing Math"].map("{:.6f}".format)
district_summary_df["% Passing Reading"] = district_summary_df["% Passing Reading"].map("{:.6f}".format)
district_summary_df["% Overall Passing"] = district_summary_df["% Overall Passing"].map("{:.6f}".format)
district_summary_df

Total Schools Total Students    Total Budget Average Math Score  \
0             15         39,170  $24,649,428.00          78.985371   

  Average Reading Score % Passing Math % Passing Reading % Overall Passing  
0              81.87784      74.980853         85.805463         65.172326

## School Summary

* Create an overview table that summarizes key metrics about each school, including:
  * School Name
  * School Type
  * Total Students
  * Total School Budget
  * Per Student Budget
  * Average Math Score
  * Average Reading Score
  * % Passing Math
  * % Passing Reading
  * % Overall Passing (The percentage of students that passed math **and** reading.)
  
* Create a dataframe to hold the above results

In [12]:
# Groupby school_name
by_school = school_data_complete.groupby(["school_name"])

# School Type
school_type = by_school["type"].first()

# Total Students
tot_students = by_school.size()

# Total School Budget
tot_school_budget = by_school["budget"].first()

# Per Student Budget
tot_per_stu = tot_school_budget/tot_students

# Average Math Score
avg_math_score = by_school["math_score"].mean()

# Average Reading Score
avg_reading_score = by_school["reading_score"].mean()

# 1 pull all students greater than or equal to 70 for math_score
# Count the number of students, and group by school name, & divide by total population of students
passing_math2 = school_data_complete[(school_data_complete["math_score"]>=70)]
percent_math2 = (passing_math2.groupby(["school_name"]).count()['student_name'] / tot_students)*100

# 1 pull all students greater than or equal to 70 for reading_score
# Count the number of students, and group by school name, & divide by total population of students
passing_reading2 = school_data_complete[(school_data_complete["reading_score"]>=70)]
percent_reading2 = (passing_reading2.groupby(["school_name"]).count()['student_name']/ tot_students)*100

# 1 pull all students greater than or equal to 70 for both math and reading
# Count the number of students, and group by school name, & divide by total population of students
passing_mathread = school_data_complete[(school_data_complete["math_score"]>=70)&(school_data_complete["reading_score"]>=70)]
percent_mathread = (passing_mathread.groupby(["school_name"]).count()['student_name']/ tot_students)*100

In [14]:
# Created School Summary DataFrame
school_summary_df = pd.DataFrame({
                                    "School Type":school_type,
                                    "Total Students":tot_students,
                                    "Total School Budget":tot_school_budget,
                                    "Per Student Budget":tot_per_stu,
                                    "Average Math Score":avg_math_score,
                                    "Average Reading Score":avg_reading_score,
                                    "% Passing Math":percent_math2,
                                    "% Passing Reading":percent_reading2,
                                    "% Passing Overall": percent_mathread
    
                                 })

In [15]:
# Cleaning up formatting with .map
school_summary_df["Total School Budget"] = school_summary_df["Total School Budget"].map("${:,.2f}".format)
school_summary_df["Per Student Budget"] = school_summary_df["Per Student Budget"].map("${:.2f}".format)
school_summary_df

School Type  Total Students Total School Budget  \
school_name                                                             
Bailey High School       District            4976       $3,124,928.00   
Cabrera High School       Charter            1858       $1,081,356.00   
Figueroa High School     District            2949       $1,884,411.00   
Ford High School         District            2739       $1,763,916.00   
Griffin High School       Charter            1468         $917,500.00   
Hernandez High School    District            4635       $3,022,020.00   
Holden High School        Charter             427         $248,087.00   
Huang High School        District            2917       $1,910,635.00   
Johnson High School      District            4761       $3,094,650.00   
Pena High School          Charter             962         $585,858.00   
Rodriguez High School    District            3999       $2,547,363.00   
Shelton High School       Charter            1761       $1,056,600.00   
Thomas High School        Charter            1635       $1,043,130.00   
Wilson High School        Charter            2283       $1,319,574.00   
Wright High School        Charter            1800       $1,049,400.00   

                      Per Student Budget  Average Math Score  \
school_name                                                    
Bailey High School               $628.00           77.048432   
Cabrera High School              $582.00           83.061895   
Figueroa High School             $639.00           76.711767   
Ford High School                 $644.00           77.102592   
Griffin High School              $625.00           83.351499   
Hernandez High School            $652.00           77.289752   
Holden High School               $581.00           83.803279   
Huang High School                $655.00           76.629414   
Johnson High School              $650.00           77.072464   
Pena High School                 $609.00           83.839917   
Rodriguez High School            $637.00           76.842711   
Shelton High School              $600.00           83.359455   
Thomas High School               $638.00           83.418349   
Wilson High School               $578.00           83.274201   
Wright High School               $583.00           83.682222   

                       Average Reading Score  % Passing Math  \
school_name                                                    
Bailey High School                 81.033963       66.680064   
Cabrera High School                83.975780       94.133477   
Figueroa High School               81.158020       65.988471   
Ford High School                   80.746258       68.309602   
Griffin High School                83.816757       93.392371   
Hernandez High School              80.934412       66.752967   
Holden High School                 83.814988       92.505855   
Huang High School                  81.182722       65.683922   
Johnson High School                80.966394       66.057551   
Pena High School                   84.044699       94.594595   
Rodriguez High School              80.744686       66.366592   
Shelton High School                83.725724       93.867121   
Thomas High School                 83.848930       93.272171   
Wilson High School                 83.989488       93.867718   
Wright High School                 83.955000       93.333333   

                       % Passing Reading  % Passing Overall  
school_name                                                  
Bailey High School             81.933280          54.642283  
Cabrera High School            97.039828          91.334769  
Figueroa High School           80.739234          53.204476  
Ford High School               79.299014          54.289887  
Griffin High School            97.138965          90.599455  
Hernandez High School          80.862999          53.527508  
Holden High School             96.252927          89.227166  
Huang High School              81.316421    

## Top Performing Schools (By % Overall Passing)

* Sort and display the top five performing schools by % overall passing.

In [ ]:
# Using sort_values, display top 5 performing schools b 5 overall passing
top_sort_overall = school_summary_df.sort_values(["% Overall Passing"],ascending = False)
top_sort_overall.head()

## Bottom Performing Schools (By % Overall Passing)

* Sort and display the five worst-performing schools by % overall passing.

In [ ]:
# Using sort_values, display 5 worst performing schools by % overall passing
worst_sort_overall = school_summary_df.sort_values(["% Overall Passing"],ascending = True)
worst_sort_overall.head()

## Math Scores by Grade

* Create a table that lists the average Reading Score for students of each grade level (9th, 10th, 11th, 12th) at each school.

  * Create a pandas series for each grade. Hint: use a conditional statement.
  
  * Group each series by school
  
  * Combine the series into a dataframe
  
  * Optional: give the displayed data cleaner formatting

In [ ]:
# Create a panda series for ninth grade using a conditional statement
# Groupby school_name and find the average math_score
ninth = school_data_complete[(school_data_complete["grade"]=='9th')]
ninth_grade = ninth.groupby(["school_name"]).mean()["math_score"]

# Create a panda series for tenth grade using a conditional statement
# Groupby school_name and find the average math_score
tenth = school_data_complete[(school_data_complete["grade"]=='10th')]
tenth_grade = tenth.groupby(["school_name"]).mean()["math_score"]

# Create a panda series for eleventh grade using a conditional statement
# Groupby school_name and find the average math_score
eleventh = school_data_complete[(school_data_complete["grade"]=='11th')]
eleventh_grade = eleventh.groupby(["school_name"]).mean()["math_score"]

# Create a panda series for twelfth grade using a conditional statement
# Groupby school_name and find the average math_score
twelfth = school_data_complete[(school_data_complete["grade"]=='12th')]
twelfth_grade = twelfth.groupby(["school_name"]).mean()["math_score"]

In [ ]:
# Create a new dataframe to hold the series data
math_scores_bygrade_df = pd.DataFrame({
                                        "9th":ninth_grade,
                                        "10th":tenth_grade,
                                        "11th":eleventh_grade,
                                        "12th":twelfth_grade
                                      })
math_scores_bygrade_df

## Reading Score by Grade 

* Perform the same operations as above for reading scores

In [ ]:
# Create a panda series for ninth grade using a conditional statement
# Groupby school_name and find the average reading_score
ninth_r = school_data_complete[(school_data_complete["grade"]=='9th')]
ninth_grade_r = ninth_r.groupby(["school_name"]).mean()["reading_score"] 

# Create a panda series for tenth grade using a conditional statement
# Groupby school_name and find the average reading_score
tenth_r = school_data_complete[(school_data_complete["grade"]=='10th')]
tenth_grade_r = tenth_r.groupby(["school_name"]).mean()["reading_score"] 

# Create a panda series for eleventh grade using a conditional statement
# Groupby school_name and find the average reading_score
eleventh_r = school_data_complete[(school_data_complete["grade"]=='11th')]
eleventh_grade_r = eleventh_r.groupby(["school_name"]).mean()["reading_score"] 

# Create a panda series for twelfth grade using a conditional statement
# Groupby school_name and find the average reading_score
twelfth_r = school_data_complete[(school_data_complete["grade"]=='12th')]
twelfth_grade_r = twelfth_r.groupby(["school_name"]).mean()["reading_score"] 

In [ ]:
reading_scores_bygrade_df = pd.DataFrame({
                                        "9th":ninth_grade_r,
                                        "10th":tenth_grade_r,
                                        "11th":eleventh_grade_r,
                                        "12th":twelfth_grade_r
                                        })
reading_scores_bygrade_df

## Scores by School Spending

* Create a table that breaks down school performances based on average Spending Ranges (Per Student). Use 4 reasonable bins to group school spending. Include in the table each of the following:
  * Average Math Score
  * Average Reading Score
  * % Passing Math
  * % Passing Reading
  * Overall Passing Rate (Average of the above two)

In [ ]:
# Make bins 
# Make group_names for labels
spending_ranges_bins = [0,585,630,645,675]
group_names = ["<$584","$585-629","$630-644", "$645-675"]

# Categorize the data frame using bins
scores_by_school = school_summary_df
scores_by_school["Spending Ranges (Per Student)"] = pd.cut(tot_per_stu, spending_ranges_bins, labels = group_names)

# Average Math Score
average_math = scores_by_school.groupby(["Spending Ranges (Per Student)"]).mean()["Average Math Score"]

# Average Reading Score
average_reading = scores_by_school.groupby(["Spending Ranges (Per Student)"]).mean()["Average Reading Score"]

# % Passing Math
percent_math = scores_by_school.groupby(["Spending Ranges (Per Student)"]).mean()["% Passing Math"]

# % Passing Reading
percent_reading = scores_by_school.groupby(["Spending Ranges (Per Student)"]).mean()["% Passing Reading"]

# % Overall Passing
percent_overall = scores_by_school.groupby(["Spending Ranges (Per Student)"]).mean()["% Overall Passing"]

In [ ]:
# Create DataFrame to store this new information
scores_by_spending_df = pd.DataFrame({
                                        "Average Math Score":average_math,
                                        "Average Reading Score":average_reading,
                                        "% Passing Math":percent_math,
                                        "% Passing Reading":percent_reading,
                                        "% Overall Passing":percent_overall
                                    })

In [ ]:
# Cleaning up formatting with .map
scores_by_spending_df["Average Math Score"] = scores_by_spending_df["Average Math Score"].map("{:.2f}".format)
scores_by_spending_df["Average Reading Score"] = scores_by_spending_df["Average Reading Score"].map("{:.2f}".format)
scores_by_spending_df["% Passing Math"] = scores_by_spending_df["% Passing Math"].map("{:.2f}".format)
scores_by_spending_df["% Passing Reading"] = scores_by_spending_df["% Passing Reading"].map("{:.2f}".format)
scores_by_spending_df["% Overall Passing"] = scores_by_spending_df["% Overall Passing"].map("{:.2f}".format)
scores_by_spending_df

## Scores by School Size

* Perform the same operations as above, based on school size.

In [ ]:
# Make bins 
# Make group_names for labels
size_range_bins = [0,1000,2000,5000]
s_group_names = ["Small (<1000)","Medium (1000-2000)","Large (2000-5000)"]

In [ ]:
# Categorize the data frame using bins
scores_by_size = school_summary_df
scores_by_size["Size Ranges"] = pd.cut(tot_students, size_range_bins, labels = s_group_names)

# Average Math Score
size_average_math = scores_by_size.groupby(["Size Ranges"]).mean()["Average Math Score"]

# Average Reading Score
size_average_reading = scores_by_size.groupby(["Size Ranges"]).mean()["Average Reading Score"]

# % Passing Math
size_percent_math = scores_by_size.groupby(["Size Ranges"]).mean()["% Passing Math"]

# % Passing Reading
size_percent_reading = scores_by_size.groupby(["Size Ranges"]).mean()["% Passing Reading"]

# % Overall Passing
size_percent_overall = scores_by_size.groupby(["Size Ranges"]).mean()["% Overall Passing"]

In [ ]:
# Create DataFrame to store this new information
scores_by_size_df = pd.DataFrame({
                                    "Average Math Score":size_average_math,
                                    "Average Reading Score":size_average_reading,
                                    "% Passing Math":size_percent_math,
                                    "% Passing Reading":size_percent_reading,
                                    "% Overall Passing":size_percent_overall
                                })
# Cleaning up formatting with .map
scores_by_size_df["Average Math Score"] = scores_by_size_df["Average Math Score"].map("{:.2f}".format)
scores_by_size_df["Average Reading Score"] = scores_by_size_df["Average Reading Score"].map("{:.2f}".format)
scores_by_size_df["% Passing Math"] = scores_by_size_df["% Passing Math"].map("{:.2f}".format)
scores_by_size_df["% Passing Reading"] = scores_by_size_df["% Passing Reading"].map("{:.2f}".format)
scores_by_size_df["% Overall Passing"] = scores_by_size_df["% Overall Passing"].map("{:.2f}".format)
scores_by_size_df

## Scores by School Type

* Perform the same operations as above, based on school type

In [ ]:
# Groupby school_type
by_type = school_summary_df.groupby(["School Type"])

# Average Math Score
type_avg_math_score = by_type["Average Math Score"].mean()

# Average Reading Score
type_avg_reading_score = by_type["Average Reading Score"].mean()

# % Passing Math
type_percent_math_score = by_type["% Passing Math"].mean()

# % Passing Reading
type_percent_reading_score = by_type["% Passing Reading"].mean()

# % Overall Passing
type_percent_overall_score = by_type["% Overall Passing"].mean()

In [ ]:
# Create new dataframe to store school type data
scores_by_type_df = pd.DataFrame({
                                    "Average Math Score":type_avg_math_score,
                                    "Average Reading Score":type_avg_reading_score,
                                    "% Passing Math":type_percent_math_score,
                                    "% Passing Reading":type_percent_reading_score,
                                    "% Overall Passing":type_percent_overall_score
                                })

In [ ]:
# Cleaning up formatting with .map
scores_by_type_df["Average Math Score"] = scores_by_type_df["Average Math Score"].map("{:.2f}".format)
scores_by_type_df["Average Reading Score"] = scores_by_type_df["Average Reading Score"].map("{:.2f}".format)
scores_by_type_df["% Passing Math"] = scores_by_type_df["% Passing Math"].map("{:.2f}".format)
scores_by_type_df["% Passing Reading"] = scores_by_type_df["% Passing Reading"].map("{:.2f}".format)
scores_by_type_df["% Overall Passing"] = scores_by_type_df["% Overall Passing"].map("{:.2f}".format)
scores_by_type_df

PyCitySchools Analysis:

It would seem from comparing the different schools in this study by size and spending that bigger and better spending does not necessarily equal better student performance. It was interesting to compare the top five best performing schools with the 5 bottom performing schools in one section. The top five performing schools were smaller with fewer students and less overall and per student budgets, and yet, seemed to outperform those schools with more students and a larger overall spending budget. The smaller schools had an overall 93/94% passing math,95-97% passing reading, and an overall passing percent of around 90. The five bottom performing schools were all larger schools with more students and a larger spending budget. Despite the larger spending budgets, these schools only had around a 65% passing math, 81% passing reading, and only an overall 53% passing. 

It was also interesting to find that the top five best performing schools were all charter schools. I think there would need to be more research done to understand the why of how charter outperformed district if there were reasons separate from school size and budget. I think it would be intersting to also compare class sizes per school or different teaching approaches/programs available to students in those schools that maybe contributed to a better success rate in the charter/smaller schools compared to the district/larger schools. 

In conclusion, taking what we know from schools in a general sense is that despite smaller budgets, smaller schools have a much better teacher-student ratio in the classroom compared to their larger counterparts where teachers may not be able to better oversee student success. 
